### ISSUE TITLE
C/C++ Interface Oddities

### ISSUE THREAD

**Issue by nabbelbabbel** [grimme-lab/xtb#98]
Using the C/C++ interface a few odd things have shown, I'd say some of them are bugs.
(If they should be split into separate Issues I don't mind doing that but for now I will gather them here.)

First of all, the interface seem to be working fine overall.  
As a C++ user I have to say: thanks for them!  

The issues encountered are:

### 1. GFN2_calculation Bond Orders
Using 
```c++
extern int
GFN2_calculation(const int* natoms, const int* attyp, const double* charge, 
const double* coord, const SCC_options* opt, const char* output, 
double* energy, double* grad, double* dipole,  double* q, double* dipm, 
double* qp, double* wbo);
```
The `wbo` array is only filled if the `SCC_options` `opt.prlevel` is `>=2`.
The expected behavior would be that the field is always populated, I think.

### 2. GFN1_calculation Bond Orders
Using 
```c++
extern int
GFN1_calculation(const int* natoms, const int* attyp, const double* charge, 
const int* uhf, const double* coord, const SCC_options* opt, const char* output, 
double* energy, double* grad, double* dipole, double* q, double* wbo);
```
The `wbo` array seems to be empty (zeroes) at any given time. 
However, using the executable the `wbo` file is created, hence I assume it should also work using the interface.
Again the expected behavior would be that the field is always populated, I think.

### 3. Parallel Execution of Calculations
3. Working in parallel it is not possible to have a single program run multiple calls to 
`GFN1_calculation` etc. at the same time. The calls do not seem to be threadsafe.
The errors range from thrown exceptions (the timings seem to have some static type members that throw reasonable exceptions) to wild segfaults.
Here, it would be good to note that these calls are not threadsafe and/or actually stop multiple calls vial locks before they go haywire.

### For Reproducibility
Commit: bbcf06e6e92853fe468241d6e4e7bc6cc2642fdf
Compiler: gcc-8.3
[Disclaimer: The code was build using CMake instead of the given toolchain]

**Comment by awvwgk** [comment]
Thanks for the report.

I guess the bond order problem is kind of known (at least I knew that I have to fix it at some point).

A lock would certainly help to prevent several `xtb` calls to run and use the same global variables, but I would rather prefer to purge all global variables from the code base than managing a mutex. So this is a long term goal.

Regarding the CMake build, would you consider sharing the build files? I've never used CMake myself for a project, but there is some interest in building `xtb` with something else than meson.

**Comment by nabbelbabbel** [comment]
I can make a PR for the CMake files once I have cleaned them up.

There could be a few minor changes required to the overall build process to keep things nice an tidy though.
(Only one list of the sources, the way `xtb_version.fh` is generated, etc.)

**Comment by awvwgk** [comment]
My last try to remove global variables was #38, but it failed eventually.

The concept of the API currently ensures that the global storage is setup correctly *every time*, this causes problems when to instances try to call the API at the same time. This also makes the API slower (just got an email complaining about this).

A solution would be to separate these steps and require to initialize and finalize `xtb` before using the API, which would be okay, I guess. The initialization state (uninitialized, GFN-FF, GFN0-xTB, GFN1-xTB or GFN2-xTB) could be saved to prevent using it with the wrong method.

**Comment by nabbelbabbel** [comment]
> The initialization state (uninitialized, GFN-FF, GFN0-xTB, GFN1-xTB or GFN2-xTB) could be saved to prevent using it with the wrong method.

So ten threads would then hold a state of their own each? That would work.

Given that I have not looked into the code that much, I'd ask what there actually is that is initialized and can be kept across calculations with different molecules.

I assume thats this is many of the parameters and precalculated stuff from the parameters?
Coming more from a C++ side, not having dealt with Fortran in a while, I wonder if Fortran has something along the lines of  `constexpr` (C++) ? This would mean that this particular chunk of data would be resolved at compile time, meaning this data is never reinitialized. (This means the data is then folded into the binary and that binary is then bigger, loads a bit slower and compiles a bit slower.)

**Comment by awvwgk** [comment]
I would prefer ten threads with independent states, but no. For the near future it will be something like the library having one global state, deciding which method is used, but all the threads can perform the calculation with the given method independently.

Generally Fortran is pretty good at evaluating even complex expressions at compile time, so this is not a problem at all. The real problem is the way the parametrisation is handled, this code needs some serious refactoring to make it possible to use compile time evaluation in the first place. Also I'm somewhat afraid of touching this part of the code, because it will cost me a lot of time and hardly anybody will notice it except for new bugs introduced.
